paired.go

```diff
@@ -86,64 +54,80 @@ func (p *paired) String(key string) (*String, error) {
 // Returns a boolean type AMF specified by the key. If the
 // key isn't found it returns a NotFoundError. If it is found
 // but is of the wrong type, this returns a WrongTypeError.
-func (p *paired) Boolean(key string) (*Boolean, error) {
+func (p *Paired) Bool(key string) (*Bool, error) {
 	val, err := p.Get(key)
 	if err != nil {
 		return nil, err
 	}
 
-	if cast, ok := val.(*Boolean); ok {
+	if cast, ok := val.(*Bool); ok {
 		return cast, nil
 	}
 
 	return nil, WrongTypeError
 }
 
-// Returns a number type AMF specified by the key. If the
-// key isn't found it returns a NotFoundError. If it is found
-// but is of the wrong type, this returns a WrongTypeError.
-func (p *paired) Number(key string) (*Number, error) {
-	val, err := p.Get(key)
-	if err != nil {
-		return nil, err
+// Returns an item specified by the key, or returns a NotFoundError.
+func (p *Paired) Get(key string) (AmfType, error) {
+	kb := []byte(key)
+	for _, tuple := range p.tuples {
+		if bytes.Compare(tuple.Key, kb) == 0 {
+			return tuple.Value, nil
+		}
 	}
 
-	if cast, ok := val.(*Number); ok {
-		return cast, nil
+	return nil, NotFoundError
+}
+
+// Decodes a single kv pair from the array
+func (p *Paired) decodePair(r io.Reader) error {
+	str := new(String)
+	if err := str.Decode(r); err != nil {
+		return err
 	}
 
-	return nil, WrongTypeError
+	value, err := Decode(r)
+	if err != nil {
+		return err
+	}
+
+	p.tuples = append(p.tuples, &tuple{
+		Key:   []byte(string(*str)),
+		Value: value,
+	})
+	return nil
 }
 
-// Returns an item specified by the key, or returns a NotFoundError.
-func (p *paired) Get(key string) (AmfType, error) {
-	kb := []byte(key)
-	for _, pair := range p.pairs {
-		if bytes.Compare(pair.Key, kb) == 0 {
-			return pair.Value, nil
-		}
+// Writes out the key pairs.
+func (p *Paired) writePairs(w io.Writer) (int, error) {
+	buf := new(bytes.Buffer)
+
+	for _, tuple := range p.tuples {
+		tuple.Encode(buf)
 	}
 
-	return nil, NotFoundError
+	n, err := io.Copy(w, buf)
+	return int(n), err
 }
 
-// Adds a new pair to the object.
-func (p *paired) Add(key string, value AmfType) {
-	p.pairs = append(p.pairs, newPair([]byte(key), value))
+type tuple struct {
+	Key   []byte
+	Value AmfType
 }
 
-// Returns the number of kv pairs in the object.
-func (p *paired) Size() int {
-	return len(p.pairs)
-}
+func (t *tuple) Encode(w io.Writer) (int, error) {
+	buf := new(bytes.Buffer)
 
-// Writes out the key pairs.
-func (p *paired) writePairs(wc *writeCollector) {
-	keylen := make([]byte, 2)
-	for _, pair := range p.pairs {
-		putUint16(keylen, 0, uint16(len(pair.Key)))
-		wc.Write(keylen)
-		wc.Write(pair.Key)
-		pair.Value.Encode(wc)
+	if _, err := NewString(string(t.Key)).Encode(buf); err != nil {
+		return 0, err
+	}
+	if _, err := buf.Write([]byte{t.Value.Marker()}); err != nil {
+		return 0, err
 	}
+	if _, err := t.Value.Encode(buf); err != nil {
+		return 0, err
+	}
```

Comment:
I can, definitely. Wanted to save on network-level writes, so I'm buffering here intentionally.